libsnark/zk_proof_systems/plonk/srs.tcc

```diff
@@ -106,12 +106,128 @@ usrs<ppT> plonk_usrs_derive_from_secret(
 
 template<typename ppT>
 srs<ppT> plonk_srs_derive_from_usrs(
-    const usrs<ppT> &usrs, const circuit_t<ppT> &circuit)
+    const usrs<ppT> &usrs,
+    const std::vector<std::vector<libff::Fr<ppT>>> gates_matrix,
+    const std::vector<size_t> wire_permutation,
+    const std::vector<size_t> PI_wire_index)
 {
+    using Field = libff::Fr<ppT>;
+
+    const size_t nrows = gates_matrix.size();
+    const size_t ncols = gates_matrix[0].size();
+    const std::vector<std::vector<Field>> gates_matrix_transpose =
+        plonk_gates_matrix_transpose(gates_matrix, nrows, ncols);
+
+    // the number of gates is equal to the number of columns in the transposed
+    // gates matrix
+    size_t num_gates = gates_matrix_transpose[0].size();
+    // ensure that num_gates is not 0
+    assert(num_gates > 0);
+    // ensure num_gates is power of 2
+    assert((num_gates & (num_gates - 1)) == 0);
+
+    // the number of Q-polynomials (aka selector polynomials) is equal to the
+    // number of rows in the transposed gates matrix
+    size_t num_qpolys = gates_matrix_transpose.size();
+
+    // the constraints q_L, q_R, q_O, q_M, q_C and the
+    // witness w_L, w_R, w_O are represented as polynomials in the roots of
+    // unity e.g. f_{q_L}(omega_i) = q_L[i], 0\le{i}<8
+    std::shared_ptr<libfqfft::evaluation_domain<Field>> domain =
+        libfqfft::get_evaluation_domain<Field>(num_gates);
+
+    // compute the selector polynomials (q-polynomials) from the
+    // transposed gates matrix over the Lagrange basis q_poly = \sum_i
+    // q[i] * L[i] where q[i] is a coefficient (a scalar Field
+    // element) and L[i] is a polynomial with Field coefficients
+    std::vector<polynomial<Field>> Q_polys =
+        plonk_compute_selector_polynomials<Field>(
+            num_gates, num_qpolys, gates_matrix_transpose, domain);
+
+    // An explanation of the constants k1,k2 from [GWC19], Section 8, page 26 :
+    //
+    // "We explicitly define the multiplicative subgroup H as containing the
+    // n-th roots of unity in F_p , where w (omega) is a primitive n-th root of
+    // unity and a generator of H i.e: H = {1, w, ... , w^{n-1}}. We assume
+    // that the number of gates in a circuit is no more than n. We also include
```

Comment:
I thought that the optimisation *is* the addition of the constants $k_1,k_2$. But maybe I have the wrong understanding?

Reply:
OK.  I took the "optmization" to be the use of degree-1 polynomials instead of the interpolant of ${1, 2, 3, 4, 5, ...}$.  To me, the $k1$ and $k2$ cosets are used to combine 3 permutation arguments into 1 (i.e. it could be used without the degree-1 polynomials.  I could be wrong, but this is how I read the bullet at the start of Section 8, which does not mention the cosets.  If that doesn't seem right, let's discuss offline.

Reply:
I think you are right. I was looking at an older version of the paper that doesn't have the bullets at the beginning of Sect. 8. Comment addressed.